lib/spack/spack/util/compression.py

```diff
@@ -412,161 +386,104 @@ def decompressor_for_win(extension):
-                return True
-        return False
-
-    @classmethod
-    def is_file_of_type(cls, iostream):
-        """Query byte stream for appropriate magic number
+    def matches_magic(self, stream: BinaryIO) -> bool:
+        """Returns true if the stream matches the current file type by any of its magic numbers.
+        Resets stream to original position.
 
         Args:
-            iostream: file byte stream
-
-        Returns:
-            Bool denoting whether file is of class file type
-            based on magic number
+            stream: file byte stream
         """
-        if not iostream:
-            return False
         # move to location of magic bytes
-        iostream.seek(cls.OFFSET)
-        magic_bytes = iostream.read(cls.header_size())
-        # return to beginning of file
-        iostream.seek(0)
-        if cls._bytes_check(magic_bytes):
-            return True
-        return False
+        offset = stream.tell()
+        stream.seek(self.OFFSET)
+        magic_bytes = stream.read(self.header_size())
+        stream.seek(offset)
+        return any(magic_bytes.startswith(magic) for magic in self.magic_numbers())
 
 
 class CompressedFileTypeInterface(FileTypeInterface):
     """Interface class for FileTypes that include compression information"""
 
-    compressed = True
+    def peek(self, stream: BinaryIO, num_bytes: int) -> Optional[io.BytesIO]:
+        """This method returns the first num_bytes of a decompressed stream. Returns None if no
+        builtin support for decompression."""
+        return None
 
-    @staticmethod
-    def decomp_in_memory(stream):
-        """This method decompresses and loads the first 200 or so bytes of a compressed file
-        to check for compressed archives. This does not decompress the entire file and should
-        not be used for direct expansion of archives/compressed files
-        """
-        raise NotImplementedError("Implementation by compression subclass required")
+
+def _decompressed_peek(
+    decompressed_stream: io.BufferedIOBase, stream: BinaryIO, num_bytes: int
+) -> io.BytesIO:
+    # Read the first num_bytes of the decompressed stream, do not advance the stream position.
+    pos = stream.tell()
+    data = decompressed_stream.read(num_bytes)
+    stream.seek(pos)
+    return io.BytesIO(data)
 
 
 class BZipFileType(CompressedFileTypeInterface):
     _MAGIC_NUMBER = b"\x42\x5a\x68"
     extension = "bz2"
+    name = "bzip2 compressed data"
 
-    @staticmethod
-    def name():
-        return "bzip2 compressed data"
-
-    @staticmethod
-    def decomp_in_memory(stream):
+    def peek(self, stream: BinaryIO, num_bytes: int) -> Optional[io.BytesIO]:
         if is_bz2_supported():
-            # checking for underlying archive, only decomp as many bytes
-            # as is absolutely neccesary for largest archive header (tar)
-            comp_stream = stream.read(TarFileType.OFFSET + TarFileType.header_size())
```

Comment:
@johnwparent this was a bug where you read the compressed stream (which could lead to issues when the decompressor requires a full block, or (unlikely but possible) the compressed file is larger than the uncompressed file when the uncompressed file is small and the compression metadata is large.